File: Frontend/src/app/order/models/order.ts
import {OrderItem} from './order-item';
import {OrderStatus} from './order-status';

export class Order {
  private _id: string = "";
  private _orderItems: OrderItem[] = [];
  private _orderDate: Date = new Date();
  private _totalPrice: number = 0;
  private _status: OrderStatus = OrderStatus.Pending;

  public constructor(id: string, orderItems: OrderItem[], orderDate: Date, totalPrice: number, status: OrderStatus) {
    this._id = id;
    this._orderItems = orderItems;
    this._orderDate = orderDate;
    this._totalPrice = totalPrice;
    this._status = status;
  }

  get id(): string {
    return this._id;
  }

  set id(value: string) {
    this._id = value;
  }

  get orderItems(): OrderItem[] {
    return this._orderItems;
  }

  set orderItems(value: OrderItem[]) {
    this._orderItems = value;
  }

  get orderDate(): Date {
    return this._orderDate;
  }

  set orderDate(value: Date) {
    this._orderDate = value;
  }

  get totalPrice(): number {
    return this._totalPrice;
  }

  set totalPrice(value: number) {
    this._totalPrice = value;
  }

  get status(): OrderStatus {
    return this._status;
  }

  set status(value: OrderStatus) {
    this._status = value;
  }

  public toJSON(): any {
    return {
      orderItems: this._orderItems.map(item => item.toJSON()),
      orderDate: this._orderDate.toISOString(),
      totalPrice: this._totalPrice,
      status: OrderStatus[this._status]
    };
  }
}

File: Backend/src/main/java/iprwc/backend/models/Order.java
package iprwc.backend.models;

import com.fasterxml.jackson.annotation.JsonFormat;
import iprwc.backend.models.enums.OrderStatus;
import jakarta.persistence.*;
import org.hibernate.annotations.JdbcTypeCode;

import java.sql.Timestamp;
import java.sql.Types;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "orders")
public class Order {
    @jakarta.persistence.Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @JdbcTypeCode(Types.VARCHAR)
    private UUID id;

    @ManyToOne
    @JoinColumn(name = "account_id", nullable = false)
    private Account account;

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<OrderItem> orderItems;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ssX")
    private Timestamp orderDate;

    private double totalPrice;

    private OrderStatus status;

    public Order() { }

    public Order(Account account, List<OrderItem> orderItems, Timestamp orderDate, double totalPrice, OrderStatus status) {
        this.account = account;
        this.orderItems = orderItems;
        this.orderDate = orderDate;
        this.totalPrice = totalPrice;
        this.status = status;
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public Account getAccount() {
        return account;
    }

    public void setAccount(Account account) {
        this.account = account;
    }

    public List<OrderItem> getOrderItems() {
        return orderItems;
    }

    public void setOrderItems(List<OrderItem> orderItems) {
        this.orderItems = orderItems;
    }

    public Timestamp getOrderDate() {
        return orderDate;
    }

    public void setOrderDate(Timestamp orderDate) {
        this.orderDate = orderDate;
    }

    public double getTotalPrice() {
        return totalPrice;
    }

    public void setTotalPrice(double totalPrice) {
        this.totalPrice = totalPrice;
    }

    public OrderStatus getStatus() {
        return status;
    }

    public void setStatus(OrderStatus status) {
        this.status = status;
    }

    @Override
    public String toString() {
        return "Order{" +
                "id=" + id +
                ", account=" + account +
                ", orderItems=" + orderItems +
                ", orderDate=" + orderDate +
                ", totalPrice=" + totalPrice +
                ", status=" + status +
                '}';
    }
}
